src/System.Linq/tests/SelectTests.cs

```diff
@@ -1118,5 +1118,42 @@ public void Select_SourceIsListSkipTakeToList()
             Assert.Equal(new[] { 6, 8 }, new List<int> { 1, 2, 3, 4 }.Select(i => i * 2).Skip(2).ToList());
             Assert.Empty(new List<int> { 1, 2, 3, 4 }.Select(i => i * 2).Skip(8).ToList());
         }
+
+        [Theory]
+        [MemberData(nameof(MoveNextAfterDisposeData))]
+        public void MoveNextAfterDispose(IEnumerable<int> source)
+        {
+            // Select is specialized for a bunch of different types, so we want
+            // to make sure this holds true for all of them.
+            var identityTransforms = new List<Func<IEnumerable<int>, IEnumerable<int>>>
+            {
+                e => e,
+                e => ForceNotCollection(e),
+                e => e.ToArray(),
+                e => e.ToList(),
+                e => new LinkedList<int>(e), // IList<T> that's not a List
+                e => e.Select(i => i) // Multiple Select() chains are optimized
+            };
+
+            foreach (IEnumerable<int> equivalentSource in identityTransforms.Select(t => t(source)))
+            {
+                IEnumerable<int> result = equivalentSource.Select(i => i);
+                using (IEnumerator<int> e = result.GetEnumerator())
+                {
+                    while (e.MoveNext()) ; // Loop until we reach the end of the iterator, @ which pt it gets disposed.
+                    e.MoveNext(); // MoveNext should not throw an exception after Dispose.
+                }
+            }
+        }
+
+        public static TheoryData<IEnumerable<int>> MoveNextAfterDisposeData()
+        {
+            return new TheoryData<IEnumerable<int>>
```

Comment:
I like TheoryData a bit better because it guarantees type/arity safety, e.g. you can't pass in the wrong type or the wrong number of arguments.